docs/implementation_design/event_driven_model.md

```diff
@@ -1 +1,73 @@
+In case of GoShimmer there is no GUI, however it applies this architecture approach as it's really flexible and is used to handle communication with other nodes and other internal parts. 
+In GoShimmer some those events can be e.g. arrival of new tangle message, peering request or plugin start. 
+When an event is triggered, an event handler (or a collection of handlers) is executed and state of the application is updated as necessary.
+ 
+## Glossary:
+
+At first let's define some terms used further to avoid misunderstandings:
+
+* Event - represents the type of event (e.g. new message or peering request) as well as set of handlers and trigger functions. Each type of event is separately defined 
+  which means that events are independent of each other - each event has its own set of handlers and is triggered separately.
+
+* Event handler (callback) - is a function that is executed when an event of given type occurs. Event handler can accept multiple arguments (e.g. message ID or plugin) so that it can perform appropriate actions.
+  Every handler must accept the same set of parameters. Each event has different set of handlers (there can be multiple handlers) that are executed when event is triggered.
+
+* Trigger - is a method that triggers execution of event handlers with given parameter values.
+
+
+## Creating new event with custom callbacks
+
+Below are the steps that show the example code necessary to create a custom event, attach a handler and trigger the event. 
+
+1. Create a function that will call event handlers (handler caller) for a specific event. 
+   Each event has only one handler caller. It that all handlers for the event must share the same interface, because the caller will pass a fixed set of arguments of specific types to handler function. 
+   It's not possible to pass different number of arguments or types to handler function. 
+   Callers for all events must also share the same interface - the first argument represents handler function that will be called represented by generic argument.
+   Further arguments represent parameters that will be passed to handler during execution. Below are example callers that accept one and two parameters respectively. 
+   More arguments can be passed in similar manner. 
+   
+```go
+func singleArgCaller(handler interface{}, params ...interface{}) {
+    handler.(func (*Plugin))(params[0].(*Plugin))
+}
+
+func twoArgsCaller(handler interface{}, params ...interface{}) {
+    handler.(func(*peer.Peer, error))(params[0].(*peer.Peer), params[1].(error))
+}
+```
+
+`handler.(func (*Plugin))(params[0].(*Plugin))` - this code seems a little complicated, so to make things simpler we will divide into smaller parts and explain each:
+
+* `handler.(func (*Plugin))` (A) - this part does type casting of handler onto type of desired function - in this case it's function that accepts `*Plugin` as its only parameters.
```

Comment:
Sounds strange to me. Maybe "this part does type cast the handler into the type of the desired function" is better? Somehow still sounds strange in my head ;)
And there is an `a` missing:
"in this case it's **a** function that accepts"

Reply:
I had a problem with this one as well. Changed this a little based on your suggestion.